# 4.8: Grouping Data & Aggregating Variables

### This script contains the following points:
#### 1. Check the dimensions of the imported dataframe
#### 2. Exercise Q2. The aggregated mean of the "order_number" column grouped by "department_id" for the entire dataframe
#### 3. Exercise Q3: How the entire dataframe results differ from the subset results
#### 4. Exercise Q4a: Loyalty flag for existing customers using the transform() function
#### 5. Exercise Q4b: Loyalty flag for existing customers using the loc() function
#### 6. Exercise Q5: User's purchased product price basic statistics
#### 7. Exercise Q6a: Spending flag for existing customers using the transform() function
#### 8. Exercise Q6b: Spending flag for existing customers using the loc() function
#### 9. Exercise Q7a: Frequency flag for existing customers using the transform() function
#### 10. Exercise Q7b: Frequency flag for existing customers using the loc() function
#### 11. Exercise Q8-9: Export dataframe as pkl file

In [54]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [55]:
# Path to main project folder

path = r'C:\Users\Mark\_Instacart Basket Analysis'

In [56]:
# Retrieve the orders_products_merged.pkl file

orders_products_merged = pd.read_pickle(os.path.join(path, '03 Scripts', 'Prepared Data', 'orders_products_merged_v2.pkl'))

#### 1. Check the dimensions of the imported dataframe

In [57]:
# Ensure nothing looks out of place with the imported dataframes

orders_products_merged.head()

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0             1   2398795        1    prior             2          3   
1             2    473747        1    prior             3          3   
2             3   2254736        1    prior             4          4   
3             4    431534        1    prior             5          4   
4             5   3367565        1    prior             6          2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  7                    15.0         196                  1   
1                 12                    21.0         196                  1   
2                  7                    29.0         196                  1   
3                 15                    28.0         196                  1   
4                  7                    19.0         196                  1   

   ...  _merge Unnamed: 0_y  product_name aisle_id  department_id  prices  \
0  ...    both          195          Soda       77              7     9.0   
1  ...    both          195          Soda       77              7     9.0   
2  ...    both          195          Soda       77              7     9.0   
3  ...    both          195          Soda       77              7     9.0   
4  ...    both          195          Soda       77              7     9.0   

   match    price_range_loc     busiest day busiest_period_of_day  
0   both  Mid-range product      Least busy        Average orders  
1   both  Mid-range product      Least busy        Average orders  
2   both  Mid-range product      Least busy        Average orders  
3   both  Mid-range product      Least busy        Average orders  
4   both  Mid-range product  Regularly busy        Average orders  

[5 rows x 21 columns]

In [58]:
# See if the data set is large, small, wide, or long

orders_products_merged.shape

(30328763, 21)

#### 2. Exercise Q2. The aggregated mean of the "order_number" column grouped by "department_id" for the entire dataframe

In [59]:
# Finding the mean with agg()
# Find the aggregated mean of the "order_number" column grouped by "department_id"

orders_products_merged.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                16.559358
2                18.413176
3                18.279600
4                18.915890
5                16.497751
6                17.609390
7                18.303975
8                16.383301
9                17.022963
10               21.227447
11               17.311768
12               16.953613
13               17.727470
14               17.898097
15               17.290123
16               18.757796
17               16.808683
18               20.336389
19               18.258620
20               17.578959
21               23.644477

In [60]:
# Using the mean() function with Square Brackets

orders_products_merged.groupby('department_id')['order_number'].mean()

department_id
1     16.559358
2     18.413176
3     18.279600
4     18.915890
5     16.497751
6     17.609390
7     18.303975
8     16.383301
9     17.022963
10    21.227447
11    17.311768
12    16.953613
13    17.727470
14    17.898097
15    17.290123
16    18.757796
17    16.808683
18    20.336389
19    18.258620
20    17.578959
21    23.644477
Name: order_number, dtype: float64

In [61]:
# Using the mean() function with Dot Notation

orders_products_merged.groupby('department_id').order_number.mean()

department_id
1     16.559358
2     18.413176
3     18.279600
4     18.915890
5     16.497751
6     17.609390
7     18.303975
8     16.383301
9     17.022963
10    21.227447
11    17.311768
12    16.953613
13    17.727470
14    17.898097
15    17.290123
16    18.757796
17    16.808683
18    20.336389
19    18.258620
20    17.578959
21    23.644477
Name: order_number, dtype: float64

In [62]:
# Multiple statistics at the same time

orders_products_merged.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                16.559358   2  99
2                18.413176   2  99
3                18.279600   2  99
4                18.915890   2  99
5                16.497751   2  99
6                17.609390   2  99
7                18.303975   2  99
8                16.383301   2  99
9                17.022963   2  99
10               21.227447   2  99
11               17.311768   2  99
12               16.953613   2  99
13               17.727470   2  99
14               17.898097   2  99
15               17.290123   2  99
16               18.757796   2  99
17               16.808683   2  99
18               20.336389   2  99
19               18.258620   2  99
20               17.578959   2  99
21               23.644477   2  99

#### Exercise Q3: How the entire dataframe results differ from the subset results: The first thing I noticed was that the number of department_ids in the subset were fewer than that of the main dataframe. This of course means that the statistics like average would differ between the two sets. For example, in the subset, the department_id 4 has a mean of 19.77, while for the same department_id 4 the mean is 18.92 in the main dataframe. The minimum and maximum in both are mostly the same except for department_id 17 of the subset, which has a maximum of 98. It is clear that a subset's statistics can be different from the main dataframe.


#### 4. Exercise Q4a: Loyalty flag for existing customers using the transform() function

In [63]:
# Step 1 (Group): the orders_products_merged dataframe is grouped by the "user_id" column
# Step 2 (Transform): the transform() function is applied to the "order_number" column with the np.max argument
# Step 3 (Store in new column): the maximum order number for each user is stored in a new column called "max_order"

orders_products_merged['max_order'] = orders_products_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [64]:
# Review dataframe

orders_products_merged.head(100)

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0              1   2398795        1    prior             2          3   
1              2    473747        1    prior             3          3   
2              3   2254736        1    prior             4          4   
3              4    431534        1    prior             5          4   
4              5   3367565        1    prior             6          2   
5              6    550135        1    prior             7          1   
6              7   3108588        1    prior             8          1   
7              8   2295261        1    prior             9          1   
8              9   2550362        1    prior            10          4   
9            144   2968173       15    prior            15          1   
10           146   1870022       15    prior            17          2   
11           147   1911383       15    prior            18          2   
12           150   2715276       15    prior            21          1   
13           151    487368       15    prior            22          1   
14           209   2293453       19    prior             2          5   
15           212   1973799       19    prior             5          6   
16           214    532817       19    prior             7          4   
17           232   1573906       21    prior            10          3   
18           458   1593000       31    prior            10          3   
19           465   2231262       31    prior            17          3   
20           617   2580647       43    prior             6          4   
21           620   2187180       43    prior             9          4   
22           752   1318871       52    prior             2          1   
23           753   1261384       52    prior             3          1   
24           755    580568       52    prior             5          2   
25           757   2428073       52    prior             7          2   
26           759    342306       52    prior             9          1   
27           760   1520353       52    prior            10          2   
28           761   2542086       52    prior            11          2   
29           762    944694       52    prior            12          1   
30           765   1268191       52    prior            15          2   
31           767   1122089       52    prior            17          2   
32           768   1498922       52    prior            18          2   
33           769    180919       52    prior            19          3   
34           771   2695875       52    prior            21          5   
35          1023    970922       67    prior             2          1   
36          1024    596585       67    prior             3          1   
37          1025   2037087       67    prior             4          2   
38          1026   2474111       67    prior             5          3   
39          1027   1568481       67    prior             6          3   
40          1029   2048527       67    prior             8          2   
41          1030   2389031       67    prior             9          2   
42          1031   2378562       67    prior            10          1   
43          1032   1515698       67    prior            11          1   
44          1033   1928050       67    prior            12          2   
45          1034   1949274       67    prior            13          1   
46          1035    551869       67    prior            14          1   
47          1037    768572       67    prior            16          4   
48          1039    150528       67    prior            18          5   
49          1040     29758       67    prior            19          0   
50          1042   1558976       67    prior            21          1   
51          1043    863974       67    prior            22          1   
52          1044   2308399       67    prior            23          1   
53          1202   2759128       81    prior             3    

In [65]:
# To see more rows set the pandas display option to "None"

pd.options.display.max_rows = None

In [66]:
# Review dataframe again

orders_products_merged.head(100)

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0              1   2398795        1    prior             2          3   
1              2    473747        1    prior             3          3   
2              3   2254736        1    prior             4          4   
3              4    431534        1    prior             5          4   
4              5   3367565        1    prior             6          2   
5              6    550135        1    prior             7          1   
6              7   3108588        1    prior             8          1   
7              8   2295261        1    prior             9          1   
8              9   2550362        1    prior            10          4   
9            144   2968173       15    prior            15          1   
10           146   1870022       15    prior            17          2   
11           147   1911383       15    prior            18          2   
12           150   2715276       15    prior            21          1   
13           151    487368       15    prior            22          1   
14           209   2293453       19    prior             2          5   
15           212   1973799       19    prior             5          6   
16           214    532817       19    prior             7          4   
17           232   1573906       21    prior            10          3   
18           458   1593000       31    prior            10          3   
19           465   2231262       31    prior            17          3   
20           617   2580647       43    prior             6          4   
21           620   2187180       43    prior             9          4   
22           752   1318871       52    prior             2          1   
23           753   1261384       52    prior             3          1   
24           755    580568       52    prior             5          2   
25           757   2428073       52    prior             7          2   
26           759    342306       52    prior             9          1   
27           760   1520353       52    prior            10          2   
28           761   2542086       52    prior            11          2   
29           762    944694       52    prior            12          1   
30           765   1268191       52    prior            15          2   
31           767   1122089       52    prior            17          2   
32           768   1498922       52    prior            18          2   
33           769    180919       52    prior            19          3   
34           771   2695875       52    prior            21          5   
35          1023    970922       67    prior             2          1   
36          1024    596585       67    prior             3          1   
37          1025   2037087       67    prior             4          2   
38          1026   2474111       67    prior             5          3   
39          1027   1568481       67    prior             6          3   
40          1029   2048527       67    prior             8          2   
41          1030   2389031       67    prior             9          2   
42          1031   2378562       67    prior            10          1   
43          1032   1515698       67    prior            11          1   
44          1033   1928050       67    prior            12          2   
45          1034   1949274       67    prior            13          1   
46          1035    551869       67    prior            14          1   
47          1037    768572       67    prior            16          4   
48          1039    150528       67    prior            18          5   
49          1040     29758       67    prior            19          0   
50          1042   1558976       67    prior            21          1   
51          1043    863974       67    prior            22          1   
52          1044   2308399       67    prior            23          1   
53          1202   2759128       81    prior             3    

#### 5. Exercise Q4b: Loyalty flag for existing customers using the loc() function

In [67]:
# Create a flag that assigns a "loyalty" label to a user ID based on its corresponding max order value
# Max orders higher than 40 - flag as a "Loyal customer"

orders_products_merged.loc[orders_products_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [68]:
# Max orders, lower than or equal to 40, and higher than 10 - flag as a "Regular customer"

orders_products_merged.loc[(orders_products_merged['max_order'] <= 40) & (orders_products_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [69]:
# Max orders lower than or equal to 10 - flag as a "New customer"

orders_products_merged.loc[orders_products_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [70]:
# Review the frequency of the "loyalty_flag" column

orders_products_merged['loyalty_flag'].value_counts(dropna = False)

Regular customer    15081691
Loyal customer      10095381
New customer         5151691
Name: loyalty_flag, dtype: int64

In [71]:
# Review dataframe

orders_products_merged.head(100)

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0              1   2398795        1    prior             2          3   
1              2    473747        1    prior             3          3   
2              3   2254736        1    prior             4          4   
3              4    431534        1    prior             5          4   
4              5   3367565        1    prior             6          2   
5              6    550135        1    prior             7          1   
6              7   3108588        1    prior             8          1   
7              8   2295261        1    prior             9          1   
8              9   2550362        1    prior            10          4   
9            144   2968173       15    prior            15          1   
10           146   1870022       15    prior            17          2   
11           147   1911383       15    prior            18          2   
12           150   2715276       15    prior            21          1   
13           151    487368       15    prior            22          1   
14           209   2293453       19    prior             2          5   
15           212   1973799       19    prior             5          6   
16           214    532817       19    prior             7          4   
17           232   1573906       21    prior            10          3   
18           458   1593000       31    prior            10          3   
19           465   2231262       31    prior            17          3   
20           617   2580647       43    prior             6          4   
21           620   2187180       43    prior             9          4   
22           752   1318871       52    prior             2          1   
23           753   1261384       52    prior             3          1   
24           755    580568       52    prior             5          2   
25           757   2428073       52    prior             7          2   
26           759    342306       52    prior             9          1   
27           760   1520353       52    prior            10          2   
28           761   2542086       52    prior            11          2   
29           762    944694       52    prior            12          1   
30           765   1268191       52    prior            15          2   
31           767   1122089       52    prior            17          2   
32           768   1498922       52    prior            18          2   
33           769    180919       52    prior            19          3   
34           771   2695875       52    prior            21          5   
35          1023    970922       67    prior             2          1   
36          1024    596585       67    prior             3          1   
37          1025   2037087       67    prior             4          2   
38          1026   2474111       67    prior             5          3   
39          1027   1568481       67    prior             6          3   
40          1029   2048527       67    prior             8          2   
41          1030   2389031       67    prior             9          2   
42          1031   2378562       67    prior            10          1   
43          1032   1515698       67    prior            11          1   
44          1033   1928050       67    prior            12          2   
45          1034   1949274       67    prior            13          1   
46          1035    551869       67    prior            14          1   
47          1037    768572       67    prior            16          4   
48          1039    150528       67    prior            18          5   
49          1040     29758       67    prior            19          0   
50          1042   1558976       67    prior            21          1   
51          1043    863974       67    prior            22          1   
52          1044   2308399       67    prior            23          1   
53          1202   2759128       81    prior             3    

In [72]:
# Review multiple columns of interest

orders_products_merged[['user_id', 'loyalty_flag', 'order_number']].head(100)

user_id      loyalty_flag  order_number
0         1      New customer             2
1         1      New customer             3
2         1      New customer             4
3         1      New customer             5
4         1      New customer             6
5         1      New customer             7
6         1      New customer             8
7         1      New customer             9
8         1      New customer            10
9        15  Regular customer            15
10       15  Regular customer            17
11       15  Regular customer            18
12       15  Regular customer            21
13       15  Regular customer            22
14       19      New customer             2
15       19      New customer             5
16       19      New customer             7
17       21  Regular customer            10
18       31  Regular customer            10
19       31  Regular customer            17
20       43  Regular customer             6
21       43  Regular customer             9
22       52  Regular customer             2
23       52  Regular customer             3
24       52  Regular customer             5
25       52  Regular customer             7
26       52  Regular customer             9
27       52  Regular customer            10
28       52  Regular customer            11
29       52  Regular customer            12
30       52  Regular customer            15
31       52  Regular customer            17
32       52  Regular customer            18
33       52  Regular customer            19
34       52  Regular customer            21
35       67  Regular customer             2
36       67  Regular customer             3
37       67  Regular customer             4
38       67  Regular customer             5
39       67  Regular customer             6
40       67  Regular customer             8
41       67  Regular customer             9
42       67  Regular customer            10
43       67  Regular customer            11
44       67  Regular customer            12
45       67  Regular customer            13
46       67  Regular customer            14
47       67  Regular customer            16
48       67  Regular customer            18
49       67  Regular customer            19
50       67  Regular customer            21
51       67  Regular customer            22
52       67  Regular customer            23
53       81      New customer             3
54       81      New customer             5
55       82  Regular customer             3
56       82  Regular customer             4
57       82  Regular customer             6
58       82  Regular customer             9
59       82  Regular customer            10
60       82  Regular customer            15
61       82  Regular customer            18
62       82  Regular customer            19
63       98  Regular customer             3
64       98  Regular customer             7
65       98  Regular customer             9
66       98  Regular customer            11
67       98  Regular customer            12
68       98  Regular customer            13
69       98  Regular customer            14
70      109      New customer             3
71      120      New customer             3
72      185      New customer             2
73      195    Loyal customer             8
74      195    Loyal customer            19
75      195    Loyal customer            22
76      195    Loyal customer            25
77      195    Loyal customer            31
78      195    Loyal customer            34
79      195    Loyal customer            37
80      195    Loyal customer            40
81      195    Loyal customer            47
82      195    Loyal customer            52
83      195    Loyal customer            57
84      222    Loyal customer            17
85      222    Loyal customer            38
86      290    Loyal customer             9
87      290    Loyal customer            28
88      331      New customer             6
89      377      New customer             3


#### 6. Exercise Q5: User's purchased product price basic statistics

In [73]:
# Multiple statistics at the same time

orders_products_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.402162  1.0  99999.0
New customer      13.396333  1.0  99999.0
Regular customer  12.546842  1.0  99999.0

#### 7. Exercise Q6a: Spending flag for existing customers using the transform() function

In [74]:
# Step 1 (Group): the orders_products_merged dataframe is grouped by the "user_id" column
# Step 2 (Transform): the transform() function is applied to the "prices" column with the np.mean argument
# Step 3 (Store in new column): the mean prices for each user is stored in a new column called "mean_prices"

orders_products_merged['mean_prices'] = orders_products_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [75]:
# Review dataframe

orders_products_merged.head(100)

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0              1   2398795        1    prior             2          3   
1              2    473747        1    prior             3          3   
2              3   2254736        1    prior             4          4   
3              4    431534        1    prior             5          4   
4              5   3367565        1    prior             6          2   
5              6    550135        1    prior             7          1   
6              7   3108588        1    prior             8          1   
7              8   2295261        1    prior             9          1   
8              9   2550362        1    prior            10          4   
9            144   2968173       15    prior            15          1   
10           146   1870022       15    prior            17          2   
11           147   1911383       15    prior            18          2   
12           150   2715276       15    prior            21          1   
13           151    487368       15    prior            22          1   
14           209   2293453       19    prior             2          5   
15           212   1973799       19    prior             5          6   
16           214    532817       19    prior             7          4   
17           232   1573906       21    prior            10          3   
18           458   1593000       31    prior            10          3   
19           465   2231262       31    prior            17          3   
20           617   2580647       43    prior             6          4   
21           620   2187180       43    prior             9          4   
22           752   1318871       52    prior             2          1   
23           753   1261384       52    prior             3          1   
24           755    580568       52    prior             5          2   
25           757   2428073       52    prior             7          2   
26           759    342306       52    prior             9          1   
27           760   1520353       52    prior            10          2   
28           761   2542086       52    prior            11          2   
29           762    944694       52    prior            12          1   
30           765   1268191       52    prior            15          2   
31           767   1122089       52    prior            17          2   
32           768   1498922       52    prior            18          2   
33           769    180919       52    prior            19          3   
34           771   2695875       52    prior            21          5   
35          1023    970922       67    prior             2          1   
36          1024    596585       67    prior             3          1   
37          1025   2037087       67    prior             4          2   
38          1026   2474111       67    prior             5          3   
39          1027   1568481       67    prior             6          3   
40          1029   2048527       67    prior             8          2   
41          1030   2389031       67    prior             9          2   
42          1031   2378562       67    prior            10          1   
43          1032   1515698       67    prior            11          1   
44          1033   1928050       67    prior            12          2   
45          1034   1949274       67    prior            13          1   
46          1035    551869       67    prior            14          1   
47          1037    768572       67    prior            16          4   
48          1039    150528       67    prior            18          5   
49          1040     29758       67    prior            19          0   
50          1042   1558976       67    prior            21          1   
51          1043    863974       67    prior            22          1   
52          1044   2308399       67    prior            23          1   
53          1202   2759128       81    prior             3    

#### 8. Exercise Q6b: Spending flag for existing customers using the loc() function

In [76]:
# Create a spending flag for each user based on the mean (average) price across all their orders
# Mean price higher than or equal to 10 - flag as a "High spender"

orders_products_merged.loc[orders_products_merged['mean_prices'] >= 10, 'spend_flag'] = 'High spender'

In [77]:
# Mean price lower than 10 - flag as a "Low spender"

orders_products_merged.loc[orders_products_merged['mean_prices'] < 10, 'spend_flag'] = 'Low spender'

In [78]:
# Review the frequency of the "spend_flag" column

orders_products_merged['spend_flag'].value_counts(dropna = False)

Low spender     29730531
High spender      598232
Name: spend_flag, dtype: int64

In [79]:
# Review dataframe

orders_products_merged.head()

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0             1   2398795        1    prior             2          3   
1             2    473747        1    prior             3          3   
2             3   2254736        1    prior             4          4   
3             4    431534        1    prior             5          4   
4             5   3367565        1    prior             6          2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  7                    15.0         196                  1   
1                 12                    21.0         196                  1   
2                  7                    29.0         196                  1   
3                 15                    28.0         196                  1   
4                  7                    19.0         196                  1   

   ...  department_id prices  match    price_range_loc     busiest day  \
0  ...              7    9.0   both  Mid-range product      Least busy   
1  ...              7    9.0   both  Mid-range product      Least busy   
2  ...              7    9.0   both  Mid-range product      Least busy   
3  ...              7    9.0   both  Mid-range product      Least busy   
4  ...              7    9.0   both  Mid-range product  Regularly busy   

   busiest_period_of_day  max_order  loyalty_flag mean_prices   spend_flag  
0         Average orders         10  New customer    6.372222  Low spender  
1         Average orders         10  New customer    6.372222  Low spender  
2         Average orders         10  New customer    6.372222  Low spender  
3         Average orders         10  New customer    6.372222  Low spender  
4         Average orders         10  New customer    6.372222  Low spender  

[5 rows x 25 columns]

In [80]:
# # Review multiple columns of interest

orders_products_merged[['user_id', 'spend_flag', 'mean_prices']].head(1000)

user_id    spend_flag  mean_prices
0          1   Low spender     6.372222
1          1   Low spender     6.372222
2          1   Low spender     6.372222
3          1   Low spender     6.372222
4          1   Low spender     6.372222
5          1   Low spender     6.372222
6          1   Low spender     6.372222
7          1   Low spender     6.372222
8          1   Low spender     6.372222
9         15   Low spender     3.994118
10        15   Low spender     3.994118
11        15   Low spender     3.994118
12        15   Low spender     3.994118
13        15   Low spender     3.994118
14        19   Low spender     8.464535
15        19   Low spender     8.464535
16        19   Low spender     8.464535
17        21   Low spender     6.987500
18        31   Low spender     8.203147
19        31   Low spender     8.203147
20        43   Low spender     7.323944
21        43   Low spender     7.323944
22        52   Low spender     6.330435
23        52   Low spender     6.330435
24        52   Low spender     6.330435
25        52   Low spender     6.330435
26        52   Low spender     6.330435
27        52   Low spender     6.330435
28        52   Low spender     6.330435
29        52   Low spender     6.330435
30        52   Low spender     6.330435
31        52   Low spender     6.330435
32        52   Low spender     6.330435
33        52   Low spender     6.330435
34        52   Low spender     6.330435
35        67   Low spender     7.635065
36        67   Low spender     7.635065
37        67   Low spender     7.635065
38        67   Low spender     7.635065
39        67   Low spender     7.635065
40        67   Low spender     7.635065
41        67   Low spender     7.635065
42        67   Low spender     7.635065
43        67   Low spender     7.635065
44        67   Low spender     7.635065
45        67   Low spender     7.635065
46        67   Low spender     7.635065
47        67   Low spender     7.635065
48        67   Low spender     7.635065
49        67   Low spender     7.635065
50        67   Low spender     7.635065
51        67   Low spender     7.635065
52        67   Low spender     7.635065
53        81   Low spender     7.800000
54        81   Low spender     7.800000
55        82   Low spender     6.270492
56        82   Low spender     6.270492
57        82   Low spender     6.270492
58        82   Low spender     6.270492
59        82   Low spender     6.270492
60        82   Low spender     6.270492
61        82   Low spender     6.270492
62        82   Low spender     6.270492
63        98   Low spender     7.763043
64        98   Low spender     7.763043
65        98   Low spender     7.763043
66        98   Low spender     7.763043
67        98   Low spender     7.763043
68        98   Low spender     7.763043
69        98   Low spender     7.763043
70       109   Low spender     6.135000
71       120  High spender    10.050000
72       185   Low spender     7.269444
73       195   Low spender     7.105464
74       195   Low spender     7.105464
75       195   Low spender     7.105464
76       195   Low spender     7.105464
77       195   Low spender     7.105464
78       195   Low spender     7.105464
79       195   Low spender     7.105464
80       195   Low spender     7.105464
81       195   Low spender     7.105464
82       195   Low spender     7.105464
83       195   Low spender     7.105464
84       222   Low spender     6.842204
85       222   Low spender     6.842204
86       290   Low spender     7.263242
87       290   Low spender     7.263242
88       331   Low spender     7.600000
89       377   Low spender     8.344444
90       387   Low spender     7.554167
91       420   Low spender     7.468966
92       420   Low spender     7.468966
93       423   Low spender     7.644444
94       423   Low spender     7.644444
95       546   Low spender     7.788889
96       591   Low spender     6.521429
97       626   Low spender     7.454193
98       626   Low spender     7.454193
99   

#### 9. Exercise Q7a: Frequency flag for existing customers using the transform() function

In [81]:
# Step 1 (Group): the orders_products_merged dataframe is grouped by the "user_id" column
# Step 2 (Transform): the transform() function is applied to the "days_since_prior_order" column with the np.mean argument
# Step 3 (Store in new column): the mean days_since_prior_order for each user is stored in a new column called "mean_days_since_prior"

orders_products_merged['mean_days_since_prior'] = orders_products_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.mean)

In [82]:
# Review dataframe

orders_products_merged.head(100)

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0              1   2398795        1    prior             2          3   
1              2    473747        1    prior             3          3   
2              3   2254736        1    prior             4          4   
3              4    431534        1    prior             5          4   
4              5   3367565        1    prior             6          2   
5              6    550135        1    prior             7          1   
6              7   3108588        1    prior             8          1   
7              8   2295261        1    prior             9          1   
8              9   2550362        1    prior            10          4   
9            144   2968173       15    prior            15          1   
10           146   1870022       15    prior            17          2   
11           147   1911383       15    prior            18          2   
12           150   2715276       15    prior            21          1   
13           151    487368       15    prior            22          1   
14           209   2293453       19    prior             2          5   
15           212   1973799       19    prior             5          6   
16           214    532817       19    prior             7          4   
17           232   1573906       21    prior            10          3   
18           458   1593000       31    prior            10          3   
19           465   2231262       31    prior            17          3   
20           617   2580647       43    prior             6          4   
21           620   2187180       43    prior             9          4   
22           752   1318871       52    prior             2          1   
23           753   1261384       52    prior             3          1   
24           755    580568       52    prior             5          2   
25           757   2428073       52    prior             7          2   
26           759    342306       52    prior             9          1   
27           760   1520353       52    prior            10          2   
28           761   2542086       52    prior            11          2   
29           762    944694       52    prior            12          1   
30           765   1268191       52    prior            15          2   
31           767   1122089       52    prior            17          2   
32           768   1498922       52    prior            18          2   
33           769    180919       52    prior            19          3   
34           771   2695875       52    prior            21          5   
35          1023    970922       67    prior             2          1   
36          1024    596585       67    prior             3          1   
37          1025   2037087       67    prior             4          2   
38          1026   2474111       67    prior             5          3   
39          1027   1568481       67    prior             6          3   
40          1029   2048527       67    prior             8          2   
41          1030   2389031       67    prior             9          2   
42          1031   2378562       67    prior            10          1   
43          1032   1515698       67    prior            11          1   
44          1033   1928050       67    prior            12          2   
45          1034   1949274       67    prior            13          1   
46          1035    551869       67    prior            14          1   
47          1037    768572       67    prior            16          4   
48          1039    150528       67    prior            18          5   
49          1040     29758       67    prior            19          0   
50          1042   1558976       67    prior            21          1   
51          1043    863974       67    prior            22          1   
52          1044   2308399       67    prior            23          1   
53          1202   2759128       81    prior             3    

#### 10. Exercise Q7b: Frequency flag for existing customers using the loc() function

In [83]:
# Create a frequency flag for each user based on the mean (average) days_since_prior_order across all their orders
# Mean days_since_prior_order higher than 20 - flag as a "Non-frequent customer"

orders_products_merged.loc[orders_products_merged['days_since_prior_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [84]:
# Mean price lower than 10 - flag as a "Frequent customer"

orders_products_merged.loc[orders_products_merged['days_since_prior_order'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [85]:
# Mean days_since_prior_order higher than 10, and lower than or equal to 20 - flag as a "Regular customer"

orders_products_merged.loc[(orders_products_merged['days_since_prior_order'] > 10) & (orders_products_merged['days_since_prior_order'] <= 20), 'frequency_flag'] = 'Regular customer'

In [86]:
# Review the frequency of the "frequency_flag" column

orders_products_merged['frequency_flag'].value_counts(dropna = False)

Frequent customer        19042466
Regular customer          6055803
Non-frequent customer     5230494
Name: frequency_flag, dtype: int64

In [87]:
# Review dataframe

orders_products_merged.head()

Unnamed: 0_x  order_id  user_id eval_set  order_number  order_dow  \
0             1   2398795        1    prior             2          3   
1             2    473747        1    prior             3          3   
2             3   2254736        1    prior             4          4   
3             4    431534        1    prior             5          4   
4             5   3367565        1    prior             6          2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  7                    15.0         196                  1   
1                 12                    21.0         196                  1   
2                  7                    29.0         196                  1   
3                 15                    28.0         196                  1   
4                  7                    19.0         196                  1   

   ...  match    price_range_loc     busiest day busiest_period_of_day  \
0  ...   both  Mid-range product      Least busy        Average orders   
1  ...   both  Mid-range product      Least busy        Average orders   
2  ...   both  Mid-range product      Least busy        Average orders   
3  ...   both  Mid-range product      Least busy        Average orders   
4  ...   both  Mid-range product  Regularly busy        Average orders   

   max_order  loyalty_flag  mean_prices   spend_flag mean_days_since_prior  \
0         10  New customer     6.372222  Low spender             20.259259   
1         10  New customer     6.372222  Low spender             20.259259   
2         10  New customer     6.372222  Low spender             20.259259   
3         10  New customer     6.372222  Low spender             20.259259   
4         10  New customer     6.372222  Low spender             20.259259   

          frequency_flag  
0       Regular customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4       Regular customer  

[5 rows x 27 columns]

In [88]:
# Review multiple columns of interest

orders_products_merged[['user_id', 'frequency_flag', 'mean_days_since_prior']].head(1000)

user_id         frequency_flag  mean_days_since_prior
0          1       Regular customer              20.259259
1          1  Non-frequent customer              20.259259
2          1  Non-frequent customer              20.259259
3          1  Non-frequent customer              20.259259
4          1       Regular customer              20.259259
5          1       Regular customer              20.259259
6          1       Regular customer              20.259259
7          1      Frequent customer              20.259259
8          1  Non-frequent customer              20.259259
9         15      Frequent customer              11.029412
10        15      Frequent customer              11.029412
11        15      Frequent customer              11.029412
12        15      Frequent customer              11.029412
13        15       Regular customer              11.029412
14        19      Frequent customer               8.040698
15        19      Frequent customer               8.040698
16        19      Frequent customer               8.040698
17        21      Frequent customer              10.055000
18        31      Frequent customer               6.073427
19        31      Frequent customer               6.073427
20        43      Frequent customer              10.549296
21        43      Frequent customer              10.549296
22        52      Frequent customer               9.155280
23        52      Frequent customer               9.155280
24        52      Frequent customer               9.155280
25        52      Frequent customer               9.155280
26        52      Frequent customer               9.155280
27        52      Frequent customer               9.155280
28        52      Frequent customer               9.155280
29        52      Frequent customer               9.155280
30        52      Frequent customer               9.155280
31        52      Frequent customer               9.155280
32        52       Regular customer               9.155280
33        52      Frequent customer               9.155280
34        52  Non-frequent customer               9.155280
35        67      Frequent customer               7.714286
36        67      Frequent customer               7.714286
37        67      Frequent customer               7.714286
38        67      Frequent customer               7.714286
39        67      Frequent customer               7.714286
40        67      Frequent customer               7.714286
41        67      Frequent customer               7.714286
42        67      Frequent customer               7.714286
43        67      Frequent customer               7.714286
44        67      Frequent customer               7.714286
45        67      Frequent customer               7.714286
46        67      Frequent customer               7.714286
47        67      Frequent customer               7.714286
48        67      Frequent customer               7.714286
49        67      Frequent customer               7.714286
50        67      Frequent customer               7.714286
51        67       Regular customer               7.714286
52        67       Regular customer               7.714286
53        81       Regular customer              26.086957
54        81  Non-frequent customer              26.086957
55        82      Frequent customer              12.188525
56        82       Regular customer              12.188525
57        82       Regular customer              12.188525
58        82  Non-frequent customer              12.188525
59        82      Frequent customer              12.188525
60        82       Regular customer              12.188525
61        82      Frequent customer              12.188525
62        82      Frequent customer              12.188525
63        98  Non-frequent customer              19.956522
64        98      Frequent customer              19.956522
65        98  Non-frequent customer              19.956522
66        98       Regular customer              19.

#### 11. Exercise Q8-9: Export dataframe as pkl file

In [89]:
# Export data to pkl

orders_products_merged.to_pickle(os.path.join(path, '03 Scripts','Prepared Data', 'orders_products_merged_v3.pkl'))

In [90]:
# Export merged data to csv

orders_products_merged.to_csv(os.path.join(path, '03 Scripts','Prepared Data', 'orders_products_merged_v3.csv'))